reconcile/openshift_namespaces.py

```diff
@@ -1,127 +1,136 @@
 
 
-QUERY = """
-{
-  namespaces: namespaces_v1 {
-    name
-    cluster {
-      name
-      serverUrl
-      jumpHost {
-          hostname
-          knownHosts
-          user
-          port
-          identity {
-              path
-              field
-              format
-          }
-      }
-      automationToken {
-        path
-        field
-        format
-      }
-      internal
-      disable {
-        integrations
-      }
-    }
-  }
-}
-"""
-
 QONTRACT_INTEGRATION = 'openshift-namespaces'
 
+NS_STATUS_PRESENT = "present"
+NS_STATUS_ABSENT = "absent"
+
+NS_ACTION_DELETE = "delete"
+NS_ACTION_CREATE = "create"
+
+
+def get_desired_state(
+        namespaces: Iterable[Dict[str, Any]]) -> List[Dict[str, str]]:
+
+    desired_state: List[Dict[str, str]] = []
+    for namespace in namespaces:
+        state = NS_STATUS_PRESENT
+        if namespace.get("delete"):
+            state = NS_STATUS_ABSENT
+
+        desired_state.append({"cluster": namespace["cluster"]["name"],
+                              "namespace": namespace["name"],
+                              "state": state})
+
+    return desired_state
 
-def get_desired_state(internal, use_jump_host, thread_pool_size):
-    gqlapi = gql.get_api()
-    all_namespaces = gqlapi.query(QUERY)['namespaces']
 
-    namespaces = []
+def get_shard_namespaces() -> List[Dict[str, str]]:
+    all_namespaces = queries.get_namespaces()
+    namespaces = {}
+    to_be_ignored = []
     for namespace in all_namespaces:
         shard_key = f'{namespace["cluster"]["name"]}/{namespace["name"]}'
         if is_in_shard(shard_key):
-            namespaces.append(namespace)
-
-    ri = ResourceInventory()
-    settings = queries.get_app_interface_settings()
-    oc_map = OC_Map(namespaces=namespaces, integration=QONTRACT_INTEGRATION,
-                    settings=settings, internal=internal,
-                    use_jump_host=use_jump_host,
-                    thread_pool_size=thread_pool_size,
-                    init_projects=True)
-    ob.init_specs_to_fetch(
-        ri,
-        oc_map,
-        namespaces=namespaces,
-        override_managed_types=['Namespace']
-    )
-
-    desired_state = []
-    for cluster, namespace, _, _ in ri:
-        if cluster not in oc_map.clusters():
-            continue
-        desired_state.append({"cluster": cluster, "namespace": namespace})
+            if shard_key not in namespaces:
+                namespaces[shard_key] = namespace
+            else:
+                to_be_ignored.append(shard_key)
+    for shard_key in to_be_ignored:
+        logging.debug(
+            f"Found multiple definitions for the namespace {shard_key};"
+            " Ignoring")
+        del namespaces[shard_key]
 
-    return oc_map, desired_state
+    return list(namespaces.values())
 
 
-def check_ns_exists(spec, oc_map):
+def check_ns_exists(spec: Mapping[str, str],
+                    oc_map: OC_Map) -> Tuple[Mapping[str, str], Any]:
```

Comment:
Just to make sure I didn't cause any confusion, you don't need to use `Mapping` for the return value, the idea is just to be as general as is possible for inputs, but we can be specific for return types.

Reply:
Why do you return the spec here? This looks like a condition that's best treated as a boolean:

```python
def ns_exists(...) -> bool:
```

and the callers do:

```python
if ns_exists(spec, oc_map):
    ...
```
I see you use it in the context of a `threaded.run`, you can pair the results with the specs with `zip`, see the comment down there.

Reply:
check the other thread as I stated before.